### PR TITLE
Not split nodes when searching for nodes but doing it all at once

### PR DESCRIPTION
**Pull request by wgliang**
**What this PR does / why we need it**:
Not split nodes when searching for nodes but doing it all at once.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
@bsalamat 

This is a follow up PR of #66733.

https://github.com/kubernetes/kubernetes/pull/66733#discussion_r205932531

**Release note**:

```release-note
Not split nodes when searching for nodes but doing it all at once.
```

### CODE REVIEW

**Comment by wgliang** [comment]
/assign @bsalamat 
PTAL

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 1]
```diff
@@ -50,3 +51,32 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
 	}
 	wg.Wait()
 }
+
+// ParallelizeUntilFeasible is a very simple framework that allow for parallelizing
+// N independent pieces of work until get feasible solution.
+func ParallelizeUntilFeasible(workers, pieces int, doWorkPiece DoWorkPieceFunc, feasible *int32) {
```
Comment (misterikkit) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: The normal way to accomplish something like this in golang would be to use a `context.Context` for the work that you are dispatching, then cancelling the context when you want to abandon the rest of the work.

It would also make sense to have the workers fan in their results to a goroutine that handles results and decides when to cancel.

**Review by misterikkit — COMMENTED** [review]
I don't think we should try too hard to mimic the `Parallelize` API that is being replaced here. If there's a pattern that works better for us here, let's go with it.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 1 above)
Comment (bsalamat) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Given that this is in client-go, I think we should make this a more generic function. More specifically, I think we should:
1. Rename it to `ParallelizeUntil`.
1. Change `feasible *int32` to a function that returns a bool and an error, similar to [ConditionFunc](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L172). When the return value is true, it stops.

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Thanks for your kind review. :). I've updated it.

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Indeed, I understand what you mean, maybe a callback function can do the same, just like bsalamat said. And using Context may be too complicated for a public function.

**Review by misterikkit — COMMENTED** [review]
> I don't think we should try too hard to mimic the Parallelize API that is being replaced here. If there's a pattern that works better for us here, let's go with it.

I apologize. I didn't notice that you were adding the util in `client-go`. Given that, it makes lots of sense to mimic the Parallelize API.

> Indeed, I understand what you mean, maybe a callback function can do the same, just like bsalamat said. And using Context may be too complicated for a public function.

I don't believe that Context should be considered "too complicated" for a public function. In this case, it offers features that we don't need, so a stop channel would also work. (Passing a stop channel around is a very common pattern in k8s, but it's an old pattern that really ought to be replaced with Context.)

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 2]
```diff
@@ -50,3 +50,36 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
...
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(workers)
+	for i := 0; i < workers; i++ {
+		go func() {
+			defer utilruntime.HandleCrash()
+			defer wg.Done()
+			for piece := range toProcess {
+				// Return and abort if the condition is satisfied.
+				if done, err := condition(); done && err == nil {
```
Comment (k82cn) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: what happen if err is NOT nil?

Comment (k82cn) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: btw, as an util, do we accept nil conditionFunc ?

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 3]
```diff
@@ -50,3 +50,36 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
...
 	wg.Wait()
 }
+
+// ConditionFunc returns true if the condition is satisfied, or an error
+// if the loop should be aborted.
+type ConditionFunc func() (done bool, err error)
+
+// ParallelizeUntil is a very simple framework that allow for parallelizing
+// N independent pieces of work until the condition function return true.
+func ParallelizeUntil(workers, pieces int, doWorkPiece DoWorkPieceFunc, condition ConditionFunc) {
```
[the remark this answers is not in the document]
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: A good idea.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 4]
```diff
@@ -50,3 +50,36 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
 	}
 	wg.Wait()
 }
+
+// ConditionFunc returns true if the condition is satisfied, or an error
+// if the loop should be aborted.
+type ConditionFunc func() (done bool, err error)
+
+// ParallelizeUntil is a very simple framework that allow for parallelizing
```
[the remark this answers is not in the document]
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: No problem.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: After this PR merge, I will submit a separate PR to replace all Parallelize.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 2 above)
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Yes,we should handle it. I'll follow misterikkit's comment and update it later.

File: pkg/scheduler/core/generic_scheduler.go [hunk 5]
```diff
@@ -413,24 +422,20 @@ func (g *genericScheduler) findNodesThatFit(pod *v1.Pod, nodes []*v1.Node) ([]*v
...
-			workqueue.Parallelize(16, int(numNodesToProcess), checkNode)
-			if filteredLen >= numNodesToFind {
-				break
-			}
-			numNodesProcessed += numNodesToProcess
-		}
+
+		// Stops searching for more nodes once the configured number of feasible nodes are found(
+		// once the remainingNodesNumber is less than or equal to 0).
+		workqueue.ParallelizeUntil(16, int(allNodes), checkNode, condition)
```
[the remark this answers is not in the document]
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: I have a question and need your guidance. If pass the channel instead of synchronization as you said, is it also needed for errs and failedPredicateMap?

[the remark this answers is not in the document]
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: And the size of the channel buffer is also a problem that needs to be determined.

[the remark this answers is not in the document]
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: Yes, I agree with you. We should pass in a channel, and then each worker function only needs data from the goroutine (here is one *v1.Node). But do we need to build this channel first? Because this may be a channel with a large size, this is my doubt.

File: pkg/scheduler/core/generic_scheduler.go [hunk 6]
```diff
@@ -413,24 +419,23 @@ func (g *genericScheduler) findNodesThatFit(pod *v1.Pod, nodes []*v1.Node) ([]*v
 			}
 			if fits {
```
Comment (bsalamat) [inline pkg/scheduler/core/generic_scheduler.go]: If we re-write this if statement as below, we can change line 377 to `filtered = make([]*v1.Node, numNodesToFind)`. This saves us some memory.

```go
if fits {
	len := atomic.AddInt32(&filteredLen, 1)
	if len > numNodesToFind {
		cancel()
	} else {
		filtered[len - 1] = g.cachedNodeInfoMap[nodeName].Node()		
	}
}
```

Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: DONE.

File: pkg/scheduler/core/generic_scheduler.go [hunk 7]
```diff
@@ -373,16 +374,19 @@ func (g *genericScheduler) findNodesThatFit(pod *v1.Pod, nodes []*v1.Node) ([]*v
 
 		// Create filtered list with enough space to avoid growing it
 		// and allow assigning.
-		filtered = make([]*v1.Node, 2*numNodesToFind)
+		filtered = make([]*v1.Node, numNodesToFind+16)
```
[the remark this answers is not in the document]
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: I'm sorry,this is what I tested and I forgot to delete it.

**Comment by wgliang** [comment]
/ping @bsalamat @k82cn
for approve

**Review by bsalamat — APPROVED** [review]
/lgtm

**Comment by bsalamat** [comment]
@kubernetes/client-go-maintainers for approval

**Comment by k82cn** [comment]
LGTM, thanks :)

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 8]
```diff
@@ -17,14 +17,15 @@ limitations under the License.
...
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 )
 
 type DoWorkPieceFunc func(piece int)
 
-// Parallelize is a very simple framework that allow for parallelizing
+// Parallelize is a very simple framework that allows for parallelizing
 // N independent pieces of work.
 func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: does it make sense to express this through the new one?

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 9]
```diff
@@ -50,3 +51,37 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
 	}
 	wg.Wait()
 }
+
+// ParallelizeUntil is a very simple framework that allows for parallelizing
+// N independent pieces of work until the context done or canceled.
+func ParallelizeUntil(workers, pieces int, doWorkPiece DoWorkPieceFunc, ctx context.Context) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: isn't ctx always the first arg by Golang convention?

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Yes, indeed. I can change it as the first ctx.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 8 above)
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: What do you think it can be? I have no idea.

Comment (sttts) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Isn't `Parallelize(...) = ParallelizeUntil(nil, ...)`?

Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Emmm, you can think of it this way, there will be a separate PR that will replace all Parallelize with ParallelizeUntil.
just like misterikkit said:https://github.com/kubernetes/kubernetes/pull/67555#discussion_r212155466

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 10]
```diff
@@ -50,3 +51,37 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
 	}
 	wg.Wait()
 }
+
+// ParallelizeUntil is a framework that allows for parallelizing N
+// independent pieces of work until the context done or canceled.
+func ParallelizeUntil(ctx context.Context, workers, pieces int, doWorkPiece DoWorkPieceFunc) {
+	// If the passed ctx is nil, then the default context.TODO() is used.
+	if ctx == nil {
+		ctx = context.TODO()
```
Comment (Huang-Wei) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: maybe change to `return Parallelize(workers, pieces, doWorkPiece)` so that we don't need to check if context is done in `select{}` for this "ctx == nil" case

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (misterikkit) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: I think there was a slight misunderstanding. We should only replace the *definition* of `Parallelize` so that existing callers won't see any change. `Parallelize` would just call `ParallelizeUntil`.

I think that is a small change that should be included in this PR.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 11]
```diff
@@ -50,3 +51,37 @@ func Parallelize(workers, pieces int, doWorkPiece DoWorkPieceFunc) {
 	}
 	wg.Wait()
 }
+
+// ParallelizeUntil is a framework that allows for parallelizing N
+// independent pieces of work until the context done or canceled.
```
Comment (misterikkit) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Suggest rewording "until the context done or canceled" to "until done or the context is canceled"

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 10 above)
Comment (misterikkit) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: Per other comments, I think the plan is for `Parallelize` to call this func with a nil context to eliminate duplicate code.

`context.TODO()` implies that we need further code changes to get an appropriate context. Since your goal is to create a context that never cancels, it is more appropriate to use `context.Background()`.

However, I think it would be cleaner to write it as,
```golang
var stop chan struct{}
if ctx != nil {
    stop = ctx.Done()
}
```

File: pkg/scheduler/core/generic_scheduler.go [hunk 12]
```diff
@@ -412,25 +416,24 @@ func (g *genericScheduler) findNodesThatFit(pod *v1.Pod, nodes []*v1.Node) ([]*v
...
-			}
-			workqueue.Parallelize(16, int(numNodesToProcess), checkNode)
-			if filteredLen >= numNodesToFind {
-				break
-			}
-			numNodesProcessed += numNodesToProcess
-		}
+
+		// Stops searching for more nodes once the configured number of feasible nodes are found(
+		// once the remainingNodesNumber is less than or equal to 0 and cancel the ctx).
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler.go]: This comment needs updating, since `remainingNodesNumber` is no longer used.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 10 above)
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: This is not a good idea, ParallelizeUntil should not rely on Parallelize.Because ParallelizeUntil may replace Parallelize.

**Review by misterikkit — CHANGES_REQUESTED** [review]
As a side note, it would be nice to rewrite the commit description and PR description to more clearly explain the change.

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 10 above)
Comment (Huang-Wei) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: That's fair.

File: pkg/scheduler/core/generic_scheduler.go (hunk 12 above)
Comment (wgliang) [reply pkg/scheduler/core/generic_scheduler.go]: Done, thanks.

**Comment by bsalamat** [comment]
/assign @sttts 

for another round of review and approval

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go (hunk 8 above)
Comment (wgliang) [reply staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: @sttts The purpose of ParallelizeUntil is to allow the caller to control through the context. This PR is already using ParallelizeUntil.
The existence of Parallelize is temporary, as other packages are still using Parallelize. After this PR, I will submit a PR delete Parallelize separately.
I don't know if you understand what I mean. :)

File: staging/src/k8s.io/client-go/util/workqueue/parallelizer.go [hunk 13]
```diff
@@ -17,34 +17,49 @@ limitations under the License.
...
+	if ctx != nil {
+		stop = ctx.Done()
+	}
+
 	toProcess := make(chan int, pieces)
 	for i := 0; i < pieces; i++ {
 		toProcess <- i
 	}
 	close(toProcess)
-
```
Comment (sttts) [inline staging/src/k8s.io/client-go/util/workqueue/parallelizer.go]: nit: would leave these lines.

**Comment by sttts** [comment]
/lgtm
/approve

For client-go changes.

**Comment by wgliang** [comment]
/ping @bsalamat
for milestone v1.12
